### PR TITLE
feat(user-interaction): support for custom events and span enhancement

### PR DESCRIPTION
**Pull request by kkruk-sumo**
## Which problem is this PR solving?

1. Only `click` events are instrumented right now. After this change, developers will have possibility to instrument any existing HTML events like `dblclick`, `touchstart`, `play` for whatever reason.
2. Created spans only have `target_element` and `target_xpath` attributes, but different websites often use custom attributes to describe elements. This change provides a callback `onSpan` which can be used to enhance created span with custom attributes, events, or even change span name. `target_xpath` is good for debugging, but when you are quickly checking out different spans, you need some more readable form of knowing which element has been clicked. Developers can also add `textContent` of clicked buttons if they want.

## Short description of the changes

Added config for the `UserInteractionInstrumentation` class with two attributes `eventTypes` and `onSpan`.

### CODE REVIEW

File: plugins/web/opentelemetry-instrumentation-user-interaction/src/instrumentation.ts
Comment (kkruk-sumo) [reply plugins/web/opentelemetry-instrumentation-user-interaction/src/instrumentation.ts]: That's a good idea! Added :)

File: plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts:33]: onSpan sounds more like an callback event approach which should return void type. I would name it like `shouldPreventSpanCreation` this way it sounds more intuitive and you can avoid adding `?` everywhere in a code by refactoring the piece in constructor - see my comment there

File: plugins/web/opentelemetry-instrumentation-user-interaction/src/instrumentation.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/src/instrumentation.ts:64]:
```suggestion

// earlier
function defaultShouldPreventSpanCreation() { return false };

....


  this._ shouldPreventSpanCreation = 
    typeof config.shouldPreventSpanCreation === 'function' ? 
      config.shouldPreventSpanCreation : 
      defaultShouldPreventSpanCreation
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:585]: whenever you use word "when" or "if" ettc. you should split the declaration with one extra level
```
describe('if event is not configured', ()=> {
  it('should not export clicks', ()=> {
    ...
  });
})
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:612]:
```
describe('when "shouldPreventSpanCreation" returns false', ()=> {
  it('should not create a span', ()=> {
    ...
  });
})
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts:253]: describe('when event was not configured to be captured', ()=> {
  it('should not export clicks', ()=> {
    ...
  });
})

File: plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts:26]: if that is possible I would create a typed event here - so that you can have type validation and allows only certain events (click, double clicks etc.)

File: plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts
Comment (kkruk-sumo) [reply plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts:33]: The main purpose of this PR was to add ability to enhance spans and IMO using `shouldPreventSpanCreation` only to add some span attributes may looks unintuitive. On the other hand I agree that `onSpan` should not expect a boolean to be returned. I was inspired by a DOM's event listeners but the 'return false' thing to prevent default behaviour is not a good pattern.
What do you think about leaving `onSpan` with the initial version (without a boolean as a return value) AND adding `shouldPreventSpanCreation` in the current form?

Comment (kkruk-sumo) [reply plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts:33]: @obecny any thoughts on that? After some consideration, single `shouldPreventSpanCreation` will also work for me.

Comment (obecny) [reply plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts:33]: I'm just trying to think of different edge cases and having `onSpan` and then `shouldPreventSpanCreation` would confuse user more than it would help. I still think that having just one function might be easier for user and also for us to keep the logic simple. I just suggested to change a name a bit so that it is easier to understand and it will have correct default behaviour (reversing the logic) if you don't specify anything

Comment (kkruk-sumo) [reply plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts]: I thought that `shouldPreventSpanCreation` may not suggest that this can be also used to enhance spans, but now I think it will be fine. Thanks :)

**Comment by kkruk-sumo** [comment]
@obecny I fixed the UTs. Are we good to merge?

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:586]:
```suggestion
    it('should not be exported not configured spans', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:615]:
```suggestion
    describe('when shouldPreventSpanCreation returns true', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:616]:
```suggestion
      it('should not record span', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.nozone.test.ts:633]:
```suggestion
  describe('when shouldPreventSpanCreation returns false', () => {
      it('should record span', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts:64]: Nit: consider changing name as it may be confused at first look with `registerInstrumentations`

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts:283]:
```suggestion
    describe('when shouldPreventSpanCreation returns true', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts:284]:
```suggestion
      it('should not record span', () => {
```

File: plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts
Comment (obecny) [inline plugins/web/opentelemetry-instrumentation-user-interaction/test/userInteraction.test.ts:301]:
```suggestion
  describe('when shouldPreventSpanCreation returns false', () => {
      it('should record span', () => {
```

**Review by obecny — APPROVED** [review]
approving, just a few nitpicks, hopefully you can address them before merging  as I will be off for next 2 weeks and don't want to block it

**Comment by kkruk-sumo** [comment]
@obecny I've fixed the failing UT's after merging main. Are we good to merge?

**Comment by kkruk-sumo** [comment]
@obecny Looks like UTs are more stable now. Previously they failed because of timeout so I don't think it was caused by this PR.

**Review by dyladan — APPROVED** [review]
LGTM when the conflict is fixed

**Review by vmarchaud — APPROVED** [review]

**Pull request merged**
